src/ProductConstructionService/ProductConstructionService.Common/RedisMutex.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Logging;
+
+namespace ProductConstructionService.Common;
+
+public interface IRedisMutex
+{
+    public Task<T> ExecuteWhenReady<T>(string mutexName, Func<Task<T>> action);
+}
+
+/// <summary>
+/// Mutex implemented using a redis cache meant for synchronization between replicas
+/// </summary>
+public class RedisMutex : IRedisMutex
+{
+    private readonly IRedisCacheFactory _cacheFactory;
+    private readonly ILogger<RedisMutex> _logger;
+
+    private const int MutexWakeUpTimeSeconds = 5;
+
+    public RedisMutex(IRedisCacheFactory cacheFactory, ILogger<RedisMutex> logger)
+    {
+        _cacheFactory = cacheFactory;
+        _logger = logger;
+    }
+
+    public async Task<T> ExecuteWhenReady<T>(string mutexName, Func<Task<T>> action)
+    {
+        IRedisCache mutexCache = _cacheFactory.Create($"{mutexName}_mutex");
+
+        try
+        {
+            // Check if a different replica is processing a subscription that belongs to the same batch
+            string? state;
+            do
+            {
+                state = await mutexCache.GetAsync();
+            } while (await Utility.SleepIfTrue(
+                () => !string.IsNullOrEmpty(state),
+                MutexWakeUpTimeSeconds,
+                () => _logger.LogInformation("Waiting for mutex {mutexName} mutexName to become available", mutexName)));
+            // Updating assets should never take more than an hour. If it does, it's possible something
+            // bad happened, so reset the mutex
```

Comment:
Again, too specific comment for a generic mutex.
The total time waiting could be an optional parameter with a default value of 1 hour to make it clear from the API

Reply:
yeah my bad I copied this part from the previous implementation, I think a parameter is a good option